enderio-base/src/main/java/com/enderio/base/common/capability/darksteel/DarkSteelUpgradeable.java

```diff
@@ -0,0 +1,161 @@
+    }
+
+    @Override
+    public void addUpgrade(IDarkSteelUpgrade upgrade) {
+        removeUpgradeInSlot(upgrade.getSlot());
+        upgrades.put(upgrade.getSerializedName(), upgrade);
+    }
+
+    @Override
+    public void removeUpgrade(String name) {
+        upgrades.remove(name);
+    }
+
+    private void removeUpgradeInSlot(String slot) {
+        Optional<String> toRemove = Optional.empty();
+        for (var entry : upgrades.entrySet()) {
+            if (entry.getValue().getSlot().equals(slot)) {
+                toRemove = Optional.of(entry.getKey());
+                break;
+            }
+        }
+        toRemove.ifPresent(upgrades::remove);
+    }
+
+    @Override
+    public boolean canApplyUpgrade(IDarkSteelUpgrade upgrade) {
+        if(upgrades.isEmpty()) {
+            return EmpoweredUpgrade.NAME.equals(upgrade.getSerializedName()) && upgrade.isBaseTier();
+        }
+
+        Optional<IDarkSteelUpgrade> existing = getUpgrade(upgrade.getSerializedName());
+        if(existing.isPresent()) {
+            return existing.get().isValidUpgrade(upgrade);
+        }
+        if(!upgrade.isBaseTier()) {
+            return false;
+        }
+        return DarkSteelUpgradeRegistry.instance().getUpgradesForItem(onItem).contains(upgrade.getSerializedName());
+    }
+
+    @Override
+    public Optional<IDarkSteelUpgrade> getUpgrade(String upgrade) {
+        return Optional.ofNullable(upgrades.get(upgrade));
+    }
+
+    @Override
+    public Collection<IDarkSteelUpgrade> getUpgrades() {
+        return upgrades.values();
+    }
+
+    @Override
+    public boolean hasUpgrade(String upgrade) {
+        return upgrades.containsKey(upgrade);
+    }
+
+    @Override
+    public Collection<IDarkSteelUpgrade> getUpgradesThatCanBeAppliedAtTheMoment() {
+        if(upgrades.isEmpty()) {
+            return List.of(EmpoweredUpgrade.TIER_0_FACTORY.get());
+        }
+        final List<IDarkSteelUpgrade> result = new ArrayList<>();
+        upgrades.values().forEach(upgrade -> upgrade.getNextTier().ifPresent(result::add));
+
+        getAllPossibleUpgrades().forEach(upgrade -> {
+            if(!hasUpgrade(upgrade.getSerializedName())) {
+                result.add(upgrade);
+            }
+        });
+        return result;
+    }
+
+    @Override
+    public Collection<IDarkSteelUpgrade> getAllPossibleUpgrades() {
+        Set<String> upgradeNames = DarkSteelUpgradeRegistry.instance().getUpgradesForItem(onItem);
+        final List<IDarkSteelUpgrade> result = new ArrayList<>();
+        upgradeNames.forEach(s -> DarkSteelUpgradeRegistry.instance().createUpgrade(s).ifPresent(result::add));
+        return result;
+    }
+
+    @Override
+    public CompoundTag serializeNBT() {
+        CompoundTag tag = new CompoundTag();
+        for (var entry : upgrades.entrySet()) {
+            tag.put(entry.getKey(), entry.getValue().serializeNBT());
+        }
+        tag.putString("onItem", onItem.toString());
```

Comment:
Might be worth considering having some sort of "constants" file or at least a constant for NBT Keys to make sure typos don't occur between reading and writing (not that there is an issue here, but is something you guys may want to discuss).

Reply:
I have create a constant in the class but I don't like the idea of a constants file. That introduces unnecessary coupling between all the things.